docs/proposals/post-deployment-hooks.md

```diff
@@ -0,0 +1,223 @@
+
+
+##### Deployment hooks
+
+Deployment hooks implemented as run-once pods external to the Kubelet provide a different set of guarantees:
+
+1. Hooks can be bound to the logical deployment lifecycle, enabling hook executions decoupled from replication mechanics.
+  1. Races can be avoided by defining a hook that executes exactly once per deployment regardless of the replica count.
+2. Hooks defined in terms of deployments are conceptually easier to reason about from the perspective of a user defining a deployment workflow.
+
+Hooks can be defined to execute before or after the deployment strategy scales up the deployment. When implementing a hook which runs after a deployment has been scaled up, there are special considerations to make:
+
+1. Nothing prevents external connectivity races: when the deployment's pods are ready, they become routable to services and other pods within an application the moment their containers enter a ready state, likely before or during or before hook execution.
+2. Hook execution can't be atomically linked to the deployment pods' statuses: If a hook failure should result in deployment failure, a previously scaled and exposed application must be rolled back when ideally the application shouldn't have been exposed prior to hook success.
+
+
+## Use cases
+
+1. As a Rails application developer, I want to perform a Rails database migration following an application deployment.
+2. As a Java application developer, I want to call a webhook following an application deployment.
+
+
+#### Use-case: Rails migrations
+
+New revisions of a Rails application often contain schema or other database updates which must accompany the new code deployment. Users should be able to specify a hook which performs a Rails migration as part of the application code deployment.
+
+Database migrations are complex and introduce downtime concerns. Here are [some examples](https://blog.rainforestqa.com/2014-06-27-zero-downtime-database-migrations) of zero-downtime Rails migration workflows.
+
+Deployments including database migrations must make special considerations:
+
+1.  Deployed code must be consistent with the currently deployed schema at any point in time.
+2.  Database or table locking must be minimized or eliminated to prevent service outages.
+
+The workflows which are effective at ensuring zero downtime migrations are typically multi-phased. For a user orchestrating a zero downtime migration deployment, it's likely the user needs to verify each deployment step discretely, with the option to abort and rollback after each phase.
+
+###### Container lifecycle hooks
+
+Container lifecycle hooks introduce problems with Rails migrations:
+
+1. The migration hook will execute once per pod in the deployment. This is a race and can result in concurrent migration attempts against a shared database. This burdens the hook author with coordinating the hooks or otherwise safeguarding the migration code to ensure it is safe for concurrent execution.
```

Comment:
I don't think that's the problem - the db is shared state for them to coordinate on.  Instead, would revise to say that this mechanism can't guarantee pods with the older code is not running.